src/librustc_mir/transform/generator.rs

```diff
@@ -0,0 +1,728 @@
+        }
+    }
+}
+
+struct TransformVisitor<'a, 'tcx: 'a> {
+    tcx: TyCtxt<'a, 'tcx, 'tcx>,
+    state_adt_ref: &'tcx AdtDef,
+    state_substs: &'tcx Substs<'tcx>,
+
+    // The index of the generator state in the generator struct
+    state_field: usize,
+
+    // Mapping from Local to (type of local, generator struct index)
+    remap: HashMap<Local, (Ty<'tcx>, usize)>,
+
+    // The number of generator states. 0 is unresumed, 1 is poisoned. So this is initialized to 2
+    bb_target_count: u32,
+
+    // Map from a (which block to resume execution at, which block to use to drop the generator)
+    // to a generator state
+    bb_targets: HashMap<(BasicBlock, Option<BasicBlock>), u32>,
+
+    // The original RETURN_POINTER local
+    new_ret_local: Local,
+
+    // The block to resume execution when for Return
+    return_block: BasicBlock,
+}
+
+impl<'a, 'tcx> TransformVisitor<'a, 'tcx> {
+    // Make a GeneratorState rvalue
+    fn make_state(&self, idx: usize, val: Operand<'tcx>) -> Rvalue<'tcx> {
+        let adt = AggregateKind::Adt(self.state_adt_ref, idx, self.state_substs, None);
+        Rvalue::Aggregate(box adt, vec![val])
+    }
+
+    // Create a Lvalue referencing a generator struct field
+    fn make_field(&self, idx: usize, ty: Ty<'tcx>) -> Lvalue<'tcx> {
+        let base = Lvalue::Local(Local::new(1));
+        let field = Projection {
+            base: base,
+            elem: ProjectionElem::Field(Field::new(idx), ty),
+        };
+        Lvalue::Projection(Box::new(field))
+    }
+
+    // Create a statement which changes the generator state
+    fn set_state(&self, state_disc: u32, source_info: SourceInfo) -> Statement<'tcx> {
+        let state = self.make_field(self.state_field, self.tcx.types.u32);
+        let val = Operand::Constant(box Constant {
+            span: source_info.span,
+            ty: self.tcx.types.u32,
+            literal: Literal::Value {
+                value: ConstVal::Integral(ConstInt::U32(state_disc)),
+            },
+        });
+        Statement {
+            source_info,
+            kind: StatementKind::Assign(state, Rvalue::Use(val)),
+        }
+    }
+}
+
+impl<'a, 'tcx> MutVisitor<'tcx> for TransformVisitor<'a, 'tcx> {
+    fn visit_lvalue(&mut self,
+                    lvalue: &mut Lvalue<'tcx>,
+                    context: LvalueContext<'tcx>,
+                    location: Location) {
+        if let Lvalue::Local(l) = *lvalue {
+            // Replace an Local in the remap with a generator struct access
+            if let Some(&(ty, idx)) = self.remap.get(&l) {
+                *lvalue = self.make_field(idx, ty);
+            }
+        } else {
+            self.super_lvalue(lvalue, context, location);
+        }
+    }
+
+    fn visit_basic_block_data(&mut self,
+                              block: BasicBlock,
+                              data: &mut BasicBlockData<'tcx>) {
+        let ret_val = match data.terminator().kind {
```

Comment:
This is *still* confusingly ordered - this block should be reordered with the `retain_statements` block.